lib/AST/Type.cpp

```diff
@@ -4566,6 +4566,41 @@ AnyFunctionType *AnyFunctionType::getAutoDiffAssociatedFunctionType(
   return associatedFunction;
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+// Compute the original function type corresponding to the given derivative
+// function type.
+AnyFunctionType *
+AnyFunctionType::getAutoDiffOriginalFunctionType() {
+  // Unwrap curry levels.
+  SmallVector<AnyFunctionType *, 2> curryLevels;
```

Comment:
I see, that makes sense. Let's fix both `getAutoDiffAssociatedFunctionType` and `getAutoDiffOriginalFunctionType` in a separate patch.

Reply:
Actually, let me commit the fix in this patch, since it's likely short.

Is it possible to do better than "always unwrap two parameter lists"?
```cpp
  auto *currentLevel = this;
  for (unsigned i : range(2)) {
    (void)i;
    if (currentLevel == nullptr)
      break;
    curryLevels.push_back(currentLevel);
    currentLevel = currentLevel->getResult()->getAs<AnyFunctionType>();
  }
```

I wonder how to detect "has self parameter" from an `AnyFunctionType`. Methods like `isInstanceMember` are defined on `FuncDecl` and aren't accessible.

Reply:
Committed code above in https://github.com/apple/swift/pull/24775/commits/f8f4e89d446949b6abcc901857fbf5b5fecbdf38.

Reply:
It's `getExtInfo().hasSelfParam()`.

Reply:
`getExtInfo().hasSelfParam()` actually returns false for methods, for some reason:
```diff
diff --git a/lib/Sema/TypeCheckAttr.cpp b/lib/Sema/TypeCheckAttr.cpp
index a1509028f7..1edf502195 100644
--- a/lib/Sema/TypeCheckAttr.cpp
+++ b/lib/Sema/TypeCheckAttr.cpp
@@ -2917,8 +2917,11 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   }

   TC.resolveDeclSignature(original);
+  auto extInfo = original->getInterfaceType()->castTo<AnyFunctionType>()->getExtInfo();
+  llvm::errs() << "ORIGINAL FUNCTION '" << original->getNameStr() << "', HAS SELF PARAM? " << extInfo.hasSelfParam() << "\n";
+  original->getInterfaceType()->dump();
   auto *originalFnTy = original->getInterfaceType()->eraseDynamicSelfType()
       ->castTo<AnyFunctionType>();
   bool isMethod = original->hasImplicitSelfDecl();
```
```test.swift
struct S {
  @differentiable(vjp: vjpFoo)
  func foo(_ x: Float) -> Float {
    return x
  }

  func vjpFoo(_ x: Float) -> (Float, (Float) -> Float) {
    return (x, { $0 })
  }
}
```
```
$ swift test.swift
ORIGINAL FUNCTION 'foo', HAS SELF PARAM? 0
(function_type escaping
  (input=function_params num_params=1
    (param
      (struct_type decl=attr.(file).S@attr.swift:1:8)))
  (output=function_type escaping
    (input=function_params num_params=1
      (param
        (struct_type decl=Swift.(file).Float)))
    (output=struct_type decl=Swift.(file).Float)))
```

I'll keep the "unwrap two parameter lists if possible" code and not pursue this further.